### ISSUE TITLE
X509: support chain building for self-signed target certs that are not directly trusted

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#19258]
It is rather untypical that the target of a certificate chain validation is a self-signed cert, 
but there are some valid cases, including directly trusted root CA or EE certs.
I recently came across a further such case, and maybe there are even more.

For supporting secure update of self-signed root CA certs ("oldWithOld")
to a new self-signed cert of the same CA ("newWithNew"),
there is the approach using a transition cert ("newWithOld")
and an optional backward transition cert ("oldWithNew"),
as described for instance in https://link.springer.com/content/pdf/10.1007/978-3-540-24852-1_20.pdf and https://www.rfc-editor.org/rfc/rfc4210#section-4.4.

[Section 6.2.  Root CA Key Update](https://www.rfc-editor.org/rfc/rfc4210#section-6.2) says:
>   CA keys (as all other keys) have a finite lifetime and will have to be updated on a periodic basis.
>   The certificates NewWithNew, NewWithOld, and OldWithNew (see Section 4.4.1) MAY be issued
>   by the CA to aid existing end entities who hold the current self-signed CA certificate (OldWithOld)
>   to transition securely to the new self-signed CA certificate (NewWithNew), and to aid new end entities
>   who will hold NewWithNew to acquire OldWithOld securely for verification of existing data.

So the idea is that the updated cert newWithNew should be verified using the old trust anchor oldWithOld and the transition cert newWithOld (and optionally vice versa).

Yet it turns out that OpenSSL does not support this.

```
openssl verify -trusted oldwithold.pem -untrusted newwithold.pem newwithnew.pem
```
yields
```
CN=Root CA
error 30 at 0 depth lookup: authority and subject key identifier mismatch
error newwithnew.pem: verification failed
```
and even the simpler
```
openssl verify -partial_chain -trusted newwithold.pem newwithnew.pem
```
does not work:
```
CN=Root CA
error 18 at 0 depth lookup: self-signed certificate
error newwithnew.pem: verification failed
```

This is because `build_chain()` implements the (AFAIK unstated) assumption that self-signed certs cannot be part of the chain and can occur as verification target only when they are directly trusted.

To still get this done I implemented a workaround with a `X509_STORE_CTX_verify_cb` function that on verification error `X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT` uses `X509_STORE_CTX_get0_chain()` to construct the required chain itself, which then can be verified as usual:

```
int selfsigned_verify_cb(int ok, X509_STORE_CTX *store_ctx)
{
    if (ok == 0 && store_ctx != NULL
        && X509_STORE_CTX_get_error_depth(store_ctx) == 0
        && X509_STORE_CTX_get_error(store_ctx)
        == X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT) {
        int i;
        STACK_OF(X509) *trusted;
        STACK_OF(X509) *chain = X509_STORE_CTX_get0_chain(store_ctx);
        STACK_OF(X509) *untrusted = X509_STORE_CTX_get0_untrusted(store_ctx);
        X509_STORE_CTX_check_issued_fn check_issued =
            X509_STORE_CTX_get_check_issued(store_ctx);
        X509 *cert = sk_X509_value(chain, 0); /* target cert */
        X509 *issuer;

        for (i = 0; i < sk_X509_num(untrusted); i++) {
            cert = sk_X509_value(untrusted, i);
            if (!X509_up_ref(cert) || !sk_X509_push(chain, cert))
                return 0;
        }

        trusted = X509_STORE_get1_all_certs(X509_STORE_CTX_get0_store(store_ctx));
        for (i = 0; i < sk_X509_num(trusted); i++) {
            issuer = sk_X509_value(trusted, i);
            if ((*check_issued)(store_ctx, cert, issuer)) {
                if (X509_up_ref(cert) && sk_X509_push(chain, cert))
                    ok = 1;
                break;
            }
        }
        sk_X509_pop_free(trusted, X509_free);
    }
    return ok;
}
```

I believe that the assumption "self-signed certs can occur as verification target only when they are directly trusted" is not really needed for chain construction.
I could provide a rather concise PR that lifts this restrictions in `build_chain()` and actually simplifies this function because some AFAICS needless portions are removed. I already convinced myself that all tests run still fine with this.

Both solutions lead the mentioned verification attempts to success:
```
openssl verify -trusted oldwithold.pem -untrusted newwithold.pem newwithnew.pem
newwithnew.pem: OK

openssl verify -partial_chain -trusted newwithold.pem newwithnew.pem
newwithnew.pem: OK
```

**Comment by DDvO** [comment]
Yep, the newWithNew cert will usually contain a different key than the oldWithOld cert (otherwise, indeed comparing the public keys to be identical would be sufficient in conjunction with checking the new self-signature).
This is also alluded by the term "Root Ca **Key** Update".

In fact the most important motivation for a root cert update is not just to extend/move the validity period but to use a fresh key pair, typically using stronger cryptography (i.e., key lengths and/or algorithms) that meanwhile is recommended/needed.

**Comment by t8m** [comment]
I think this is fundamentally different operation than the normal certificate verification. IMO it should not be convoluted like this with X509_verify_cert() and there should be a different verification API call for it. It could be seen as two operations - 1. verify the cross-signed certificate 2. verify that the new self-signed root has the same key and subject (or potentially some other attributes - path len, other constraints).

**Comment by DDvO** [comment]
> I think this is fundamentally different operation than the normal certificate verification. IMO it should not be convoluted like this with `X509_verify_cert()` and there should be a different verification API call for it.

Having a different API just for this special case would incur some extra effort and l fear quite some code duplication.

> It could be seen as two operations -
> 1. verify the cross-signed certificate

This can be achieved pretty elegantly with a minor extension of the existing API:
- either generally allowing self-signed target certs, or if folks are at unease with that,
- provide a new `X509_V_FLAG_SELFSIGNED_TARGET` that is off by default.

Both these options also have the advantage of generality: they can serve also other use cases rather than just root CA updates.

> 2. verify that the new self-signed root has the same key and subject (or potentially some other attributes - path len, other constraints).

As I wrote above, the case that the public keys are identical should not be of much relevance.
If it is considered worth supporting specifically, we could have an extra function like
```
int X509_check_selfsigned_update_same_key(const X509 *old, const X509 *new)
```
but with my proposal for part 1, this is not needed - the generalized `X509_verify_cert()` will cover this special case as well.

**Comment by DDvO** [comment]
Another solution at least for the API level would be to add a verification function that does not perform the chain building itself but relies on the caller to do it (similarly to the workaround using the verification callback function that I shared above).

This could look like
```
int X509_verify_chain_cert(X509_STORE_CTX *ctx, STACK_OF(X509) *chain)
```
where, as for `X509_verify_cert()`, the caller is also expected to provide a `ctx` that includes the target cert and optionally a trust store, which is just used for providing verification options
or
```
int X509_verify_chain_cert(X509_STORE* store, STACK_OF(X509) *chain, X509 *target)
```
where the `store` parameter is optional, just used for providing verification options.
In both flavors, the corner case `chain == NULL` (or an empty list) could be allowed.

**Comment by DDvO** [comment]
@vdukhovni, what do you think about all this?

**Comment by nhorman** [comment]
ping @vdukhovni

**Comment by vdukhovni** [comment]
My take is that bootstrapping of new trust anchors is so far removed from normal chain construction and verification that it *should not* be supported by either the normal verification API or a separate API.  

Complicating and risking issues in chain construction to support what is basically a trust-store management function makes no sense to me.

Rather, if one wants to sanity check updates to the trust store, a separate tool (CLI not API) can be implemented that can perform the "oldWithOld -> newWithNew" chaining outside of library code. Perhaps such a CLI tool should be added to the OpenSSL command-line, or it can be a separate project that uses OpenSSL or some other crypto library under the hood, but I don't see a need to burden normal x509 chain construction/validation with this entirely separate corner case.
